File: examples/python/03-functions/main.py
"""
Functions in Python
===================
Python has first-class functions: they are objects, can be passed around,
returned from other functions, and stored in data structures.
"""

from functools import reduce, partial


# ── Basic function definition ─────────────────────────

def greet(name: str) -> str:
    """A simple function with type hints and a docstring."""
    return f"Hello, {name}!"


print(greet("Atlas"))
print(f"greet is a {type(greet)}")  # <class 'function'>


# ── Default and keyword arguments ─────────────────────

def power(base: float, exponent: float = 2) -> float:
    """Default argument: exponent defaults to 2 (square)."""
    return base ** exponent


print(f"\npower(3) = {power(3)}")          # 9
print(f"power(2, 10) = {power(2, 10)}")    # 1024
print(f"power(base=5, exponent=3) = {power(base=5, exponent=3)}")  # 125


# ── *args and **kwargs ────────────────────────────────

def flexible(*args, **kwargs):
    """Accept any number of positional and keyword arguments."""
    print(f"  args = {args}")
    print(f"  kwargs = {kwargs}")


print("\nflexible(1, 2, 3, name='Ada', age=36):")
flexible(1, 2, 3, name="Ada", age=36)


# ── Functions are first-class objects ─────────────────

print("\n--- First-class functions ---")

def add(a, b):
    return a + b


def multiply(a, b):
    return a * b


# Functions can be assigned to variables
operation = add
print(f"operation(3, 4) = {operation(3, 4)}")  # 7

# Functions can be stored in data structures
ops = {"add": add, "mul": multiply}
print(f"ops['mul'](3, 4) = {ops['mul'](3, 4)}")  # 12


# ── Higher-order functions ────────────────────────────

print("\n--- Higher-order functions ---")

numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

# map — apply a function to each element
squares = list(map(lambda x: x ** 2, numbers))
print(f"squares = {squares}")

# filter — keep elements that satisfy a predicate
evens = list(filter(lambda x: x % 2 == 0, numbers))
print(f"evens = {evens}")

# reduce — combine all elements into one value
total = reduce(lambda acc, x: acc + x, numbers, 0)
print(f"sum = {total}")

product = reduce(lambda acc, x: acc * x, numbers, 1)
print(f"product = {product}")

# sorted with key function
words = ["banana", "apple", "cherry", "date"]
by_length = sorted(words, key=len)
print(f"sorted by length = {by_length}")


# ── Writing your own higher-order function ────────────

print("\n--- Custom higher-order function ---")


def apply_twice(f, x):
    """Apply function f to x, then apply f to result."""
    return f(f(x))


print(f"apply_twice(lambda x: x + 1, 5) = {apply_twice(lambda x: x + 1, 5)}")  # 7
print(f"apply_twice(lambda x: x * 2, 3) = {apply_twice(lambda x: x * 2, 3)}")  # 12


# ── Lambda (anonymous functions) ──────────────────────

print("\n--- Lambda ---")

# Lambda: single expression, no statements, no name
square = lambda x: x ** 2
print(f"square(5) = {square(5)}")

# Useful inline with higher-order functions
pairs = [(1, "one"), (3, "three"), (2, "two")]
sorted_pairs = sorted(pairs, key=lambda p: p[0])
print(f"sorted pairs = {sorted_pairs}")


# ── Closures ──────────────────────────────────────────

print("\n--- Closures ---")


def make_multiplier(factor):
    """Return a function that multiplies its argument by factor."""
    def multiplier(x):
        return x * factor  # factor is captured from enclosing scope
    return multiplier


double = make_multiplier(2)
triple = make_multiplier(3)

print(f"double(5) = {double(5)}")    # 10
print(f"triple(5) = {triple(5)}")    # 15

# The closure captures variable, not value
print(f"double is a {type(double)}")
print(f"double.__closure__[0].cell_contents = {double.__closure__[0].cell_contents}")


def make_counter(start=0):
    """Closure with mutable state (using a list as a box)."""
    count = [start]  # list because nonlocal reassignment needs Python 3

    def increment():
        count[0] += 1
        return count[0]

    return increment


counter = make_counter()
print(f"counter() = {counter()}")  # 1
print(f"counter() = {counter()}")  # 2
print(f"counter() = {counter()}")  # 3


# ── Partial application ──────────────────────────────

print("\n--- Partial application ---")


def log(level, message):
    print(f"[{level}] {message}")


# Fix first argument
info = partial(log, "INFO")
error = partial(log, "ERROR")

info("Server started")
error("Connection failed")


# ── Decorators ────────────────────────────────────────

print("\n--- Decorators ---")


def log_calls(func):
    """Decorator that logs function calls."""
    def wrapper(*args, **kwargs):
        print(f"  → Calling {func.__name__}({args}, {kwargs})")
        result = func(*args, **kwargs)
        print(f"  ← {func.__name__} returned {result}")
        return result
    return wrapper


@log_calls
def add_numbers(a, b):
    return a + b


result = add_numbers(3, 4)


def memoize(func):
    """Decorator that caches results."""
    cache = {}

    def wrapper(*args):
        if args not in cache:
            cache[args] = func(*args)
        return cache[args]

    return wrapper


@memoize
def fibonacci(n):
    if n < 2:
        return n
    return fibonacci(n - 1) + fibonacci(n - 2)


print(f"\nfibonacci(10) = {fibonacci(10)}")
print(f"fibonacci(30) = {fibonacci(30)}")  # fast thanks to memoization


# ── Generators (lazy functions) ───────────────────────

print("\n--- Generators ---")


def fibonacci_gen():
    """Generate Fibonacci numbers lazily (infinite)."""
    a, b = 0, 1
    while True:
        yield a
        a, b = b, a + b


# Take first 10 from infinite generator
from itertools import islice

fibs = list(islice(fibonacci_gen(), 10))
print(f"First 10 Fibonacci: {fibs}")

# Generator expression (like list comprehension but lazy)
squares_gen = (x ** 2 for x in range(1, 11))
print(f"Squares (generator): {list(squares_gen)}")


# ── Function composition ─────────────────────────────

print("\n--- Function composition ---")


def compose(*fns):
    """Compose functions right-to-left: compose(f, g)(x) = f(g(x))"""
    def composed(x):
        result = x
        for f in reversed(fns):
            result = f(result)
        return result
    return composed


shout = compose(str.upper, str.strip, lambda s: s + "!")
print(f'shout("  hello  ") = "{shout("  hello  ")}"')

# Pipeline style (left-to-right)
def pipe(*fns):
    """Pipe functions left-to-right: pipe(f, g)(x) = g(f(x))"""
    def piped(x):
        result = x
        for f in fns:
            result = f(result)
        return result
    return piped


process = pipe(
    str.strip,
    str.lower,
    str.title,
    lambda s: f"Hello, {s}!",
)
print(f'process("  alice  ") = "{process("  alice  ")}"')


# ── Summary ───────────────────────────────────────────

print("\n--- Summary ---")
print("Python functions:")
print("  - First-class: assignable, passable, storable")
print("  - Higher-order: map, filter, reduce, sorted(key=...)")
print("  - Lambda: anonymous single-expression functions")
print("  - Closures: capture enclosing scope")
print("  - Partial application: functools.partial")
print("  - Decorators: wrap functions with @syntax")
print("  - Generators: lazy evaluation with yield")
print("  - Composition: build pipelines of functions")

File: examples/clojure/03-functions/main.clj
;;; Functions in Clojure
;;; ====================
;;; Functions are the primary abstraction in Clojure. They are first-class,
;;; can be composed, and are building blocks of all Clojure programs.

(ns main
  (:require [clojure.string :as str]))

;; ── Basic functions ──────────────────────────────────

(defn greet
  "A simple function with a docstring."
  [name]
  (str "Hello, " name "!"))

(println (greet "Atlas"))

;; Multi-arity (overloaded by argument count)
(defn power
  "Raise base to exponent. Defaults to squaring."
  ([base] (power base 2))
  ([base exp] (Math/pow base exp)))

(println "\n--- Multi-arity ---")
(println "(power 3) =" (power 3))
(println "(power 2 10) =" (power 2 10))

;; Variadic arguments
(defn sum [& nums]
  (reduce + 0 nums))

(println "(sum 1 2 3 4 5) =" (sum 1 2 3 4 5))

;; ── First-class functions ────────────────────────────

(println "\n--- First-class functions ---")

;; Functions are values — assignable, passable, storable
(def op +)
(println "(op 3 4) =" (op 3 4))

(def ops {:add + :sub - :mul * :div /})
(println "(:mul ops) 3 4 =" ((:mul ops) 3 4))

;; ── Anonymous functions ──────────────────────────────

(println "\n--- Anonymous functions ---")

;; fn form
(println "((fn [x] (* x x)) 5) =" ((fn [x] (* x x)) 5))

;; Reader macro #()
(println "(#(* % %) 5) =" (#(* % %) 5))
(println "(#(+ %1 %2) 3 4) =" (#(+ %1 %2) 3 4))

;; ── Higher-order functions ───────────────────────────

(println "\n--- Higher-order functions ---")

(def numbers (range 1 11))

;; map
(println "(map #(* % %) numbers) =" (vec (map #(* % %) numbers)))

;; filter
(println "(filter even? numbers) =" (vec (filter even? numbers)))

;; reduce (fold)
(println "(reduce + numbers) =" (reduce + numbers))
(println "(reduce * 1 numbers) =" (reduce * 1 numbers))

;; remove (opposite of filter)
(println "(remove even? numbers) =" (vec (remove even? numbers)))

;; some, every?
(println "(some even? numbers) =" (some even? numbers))
(println "(every? pos? numbers) =" (every? pos? numbers))

;; sort-by
(def words ["banana" "apple" "cherry" "date"])
(println "(sort-by count words) =" (sort-by count words))

;; group-by
(println "(group-by even? [1..6]) ="
         (group-by even? (range 1 7)))

;; ── Function composition ────────────────────────────

(println "\n--- Composition ---")

;; comp: right-to-left composition
(def shout (comp str/upper-case str/trim))
(println "(shout \"  hello  \") =" (shout "  hello  "))

(def count-evens (comp count (partial filter even?)))
(println "(count-evens [1..20]) =" (count-evens (range 1 21)))

;; Threading macros: left-to-right pipelines
(println
  "Thread-last: "
  (->> (range 1 11)
       (filter even?)
       (map #(* % %))
       (reduce +)))
;; 1. numbers 1-10
;; 2. keep evens: 2,4,6,8,10
;; 3. square: 4,16,36,64,100
;; 4. sum: 220

;; Thread-first: for object-like operations
(println
  "Thread-first: "
  (-> "  Hello, World!  "
      str/trim
      str/lower-case
      (str/replace "world" "atlas")))

;; ── Partial application ──────────────────────────────

(println "\n--- Partial application ---")

(def add5 (partial + 5))
(println "(add5 10) =" (add5 10))

(def info-log (partial println "[INFO]"))
(info-log "Server started")

;; ── Closures ─────────────────────────────────────────

(println "\n--- Closures ---")

(defn make-multiplier [factor]
  (fn [x] (* x factor)))

(def double' (make-multiplier 2))
(def triple' (make-multiplier 3))
(println "(double' 5) =" (double' 5))
(println "(triple' 5) =" (triple' 5))

;; Counter with atom (managed mutable state)
(defn make-counter []
  (let [count (atom 0)]
    (fn [] (swap! count inc))))

(def counter (make-counter))
(println "(counter) =" (counter))
(println "(counter) =" (counter))
(println "(counter) =" (counter))

;; ── Destructuring in function args ───────────────────

(println "\n--- Destructuring ---")

;; Vector destructuring
(defn distance [[x1 y1] [x2 y2]]
  (Math/sqrt (+ (Math/pow (- x2 x1) 2)
                (Math/pow (- y2 y1) 2))))

(println "(distance [0 0] [3 4]) =" (distance [0 0] [3 4]))

;; Map destructuring
(defn greet-person [{:keys [name age]}]
  (str name " is " age " years old"))

(println (greet-person {:name "Ada" :age 36}))

;; ── Recursion ────────────────────────────────────────

(println "\n--- Recursion ---")

;; Direct recursion
(defn factorial [n]
  (if (<= n 1) 1
      (* n (factorial (dec n)))))

(println "(factorial 10) =" (factorial 10))

;; Tail recursion with recur (required for TCO in Clojure)
(defn factorial-tail [n]
  (loop [i n, acc 1]
    (if (<= i 1) acc
        (recur (dec i) (* acc i)))))

(println "(factorial-tail 20) =" (factorial-tail 20))

;; ── Lazy sequences ──────────────────────────────────

(println "\n--- Lazy sequences ---")

;; iterate: generate lazy sequence by repeated application
(println "(take 10 (iterate #(* 2 %) 1)) ="
         (vec (take 10 (iterate #(* 2 %) 1))))

;; Fibonacci as lazy sequence
(def fibs
  (map first (iterate (fn [[a b]] [b (+ a b)]) [0 1])))
(println "(take 15 fibs) =" (vec (take 15 fibs)))

;; ── Memoization ──────────────────────────────────────

(println "\n--- Memoization ---")

(def fib-memo
  (memoize
    (fn [n]
      (if (< n 2) n
          (+ (fib-memo (dec n)) (fib-memo (- n 2)))))))

(println "(fib-memo 30) =" (fib-memo 30))

;; ── apply ────────────────────────────────────────────

(println "\n--- apply ---")

;; apply "spreads" a collection as arguments
(println "(apply + [1 2 3 4 5]) =" (apply + [1 2 3 4 5]))
(println "(apply max [3 1 4 1 5 9]) =" (apply max [3 1 4 1 5 9]))
(println "(apply str [\"a\" \"b\" \"c\"]) =" (apply str ["a" "b" "c"]))

;; ── juxt (apply multiple functions at once) ──────────

(println "\n--- juxt ---")

(def stats (juxt count #(reduce + %) #(apply min %) #(apply max %)))
(println "(stats [3 1 4 1 5 9]) =" (stats [3 1 4 1 5 9]))
;; [count, sum, min, max]

;; ── Summary ──────────────────────────────────────────

(println "\n--- Summary ---")
(println "Clojure functions:")
(println "  - First-class: values, passable, composable")
(println "  - defn for named, fn / #() for anonymous")
(println "  - Multi-arity and variadic")
(println "  - Closures capture enclosing scope")
(println "  - map, filter, reduce, remove, some, every?")
(println "  - comp for composition, partial for currying")
(println "  - -> and ->> threading macros for pipelines")
(println "  - Destructuring in function parameters")
(println "  - recur for tail-call optimisation")
(println "  - Lazy sequences with iterate, lazy-seq")
(println "  - memoize for automatic caching")
